File: 52.js
function shortestPath(grid) {
  const n = grid.length, m = grid[0].length;
  if (grid[0][0] === 1) return -1;
  const q = [[0, 0, 0]];
  const visited = new Set(["0,0"]);

  const dirs = [[1,0],[-1,0],[0,1],[0,-1]];
  while (q.length) {
    const [x, y, d] = q.shift();
    if (x === n - 1 && y === m - 1) return d;
    for (const [dx, dy] of dirs) {
      const nx = x + dx, ny = y + dy;
      const key = `${nx},${ny}`;
      if (nx>=0 && ny>=0 && nx<n && ny<m && grid[nx][ny]===0 && !visited.has(key)) {
        visited.add(key);
        q.push([nx, ny, d + 1]);
      }
    }
  }
  return -1;
}

File: 52.py
from collections import deque

def shortestPath(grid):
    n, m = len(grid), len(grid[0])
    if grid[0][0] == 1: return -1
    q = deque([(0, 0, 0)])
    visited = set([(0, 0)])
    
    while q:
        x, y, d = q.popleft()
        if x == n - 1 and y == m - 1:
            return d
        for dx, dy in [(1,0),(-1,0),(0,1),(0,-1)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == 0 and (nx, ny) not in visited:
                visited.add((nx, ny))
                q.append((nx, ny, d + 1))
    return -1
